### DISCUSSION TITLE
Double-precision to single-precision float conversions in pymilvus package

### DISCUSSION THREAD

**Discussion by drassokhin** [milvus-io/milvus#20870] [Q&A and General discussion]
I noticed that the ```round``` built-in Python method is used in several places in the pymilvus package to convert from double-precision (64-bit) values to the float precision (32-bit). For example, this is done here: https://github.com/milvus-io/pymilvus/blob/master/pymilvus/client/grpc_handler.py#L921 and here: https://github.com/milvus-io/pymilvus/blob/master/pymilvus/client/abstract.py#L280. This is actually incorrect and can lead to the wrong results. A 32-bit float has 23 bits of significand, 8 bits of exponent, and 1 sign bit, and has approximately 7 decimal digit precision, and the exponent ranging from -38 to 38 (in the decimal notation). The built-in ```round``` method returns a rounded version of the specified floating-point number with the specified number of decimals after the period, without taking the exponent part into account. So, for example, ```round(1.23456789e-10, 6)``` will return 0.0 rather than 1.234567e-10, which one would expect when converting this value from the 64-bit double to 32-bit float.  

```
import numpy as np
np.double(0.00000123456789) # 64-bit double, can represent up to 16 decimal digits
Out[67]: 1.23456789e-06
np.single(0.00000123456789)  # convert to 32-bit float, will get 7 to 8 significant decimal digits
Out[68]: 1.2345679e-06
round(0.00000123456789, 6)   # will lose all significant digits but one
Out[69]: 1e-06
```

**Comment by yhmo** [comment]
Milvus server stores float vector as float32 array. All the ANN search computation is based on float32 type.
Although the python client use double-precision array to store vector data, the value in this array actually is float32, the digits after the no.6 position are meaningless. So, we don't need to return double-precision value to users, we only can ensure the digits before no.6 position is correct.

**Reply by drassokhin** [reply]
@yhmo What I am saying is that calling Python's ```round(x, 6)``` is not the same as converting a 64-bit IEEE double to a 32-bit float! For example, Python's ```round(x, 6)``` would convert 0.0000001 to exactly 0.0, whereas, using the C/C++ notation, ```(float)0.0000001``` (or ```1.0e-7f```) is still > 0.0 and can be represented by a 32-bit float just fine. You are confusing the 'positions after the decimal point' with the precision. All usages of ```round(x, 6)``` must absolutely be eliminated from the pymilvus code, as they will result in errors.

**Reply by yhmo** [reply]
@XuanYang-cn will double-check this issue.

**Comment by XuanYang-cn** [comment]
Both `round` are used to parse search and query results. Yes, using round(, 6) may lose precision, I'll pull a request to fix it.

But don't worry, the data in Milvus is correct.
